src/Storages/ObjectStorage/DataLakes/Iceberg/Mutations.cpp

```diff
@@ -203,12 +238,79 @@ std::optional<DeleteFileWriteResultWithStats> writeDataFiles(
+        settings.return_mutated_rows = true;
+
+        auto interpreter = std::make_unique<MutationsInterpreter>(storage_ptr, metadata, commands, context, settings);
+        auto pipeline = QueryPipelineBuilder::getPipeline(interpreter->execute());
+        PullingPipelineExecutor executor(pipeline);
+
+        auto header = interpreter->getUpdatedHeader();
+
+        Block block;
+        while (executor.pull(block))
+        {
+            auto data_block = getNonVirtualColumns(block);
+            Chunk chunk(data_block.getColumns(), data_block.rows());
+            auto partition_result = chunk_partitioner.partitionChunk(chunk);
+
+            for (const auto & [partition_key, partition_chunk] : partition_result)
+            {
+                if (!update_data_statistics.contains(partition_key))
+                    update_data_statistics.emplace(partition_key, DataFileStatistics(data_schema->getArray(Iceberg::f_fields)));
+
+                auto it = update_data_writers.find(partition_key);
+                if (it == update_data_writers.end())
+                {
+                    auto data_file_info = generator.generateDataFileName();
+                    update_data_result[partition_key].path = data_file_info;
+                    auto data_write_buffer = object_storage->writeObject(
+                        StoredObject(data_file_info.path_in_storage),
+                        WriteMode::Rewrite,
+                        std::nullopt,
+                        DBMS_DEFAULT_BUFFER_SIZE,
+                        context->getWriteSettings());
+
+                    auto data_output_format = FormatFactory::instance().getOutputFormat(
+                        configuration->format, *data_write_buffer, data_block, context, format_settings, nullptr);
+
+                    update_data_write_buffers[partition_key] = std::move(data_write_buffer);
+                    it = update_data_writers.emplace(partition_key, std::move(data_output_format)).first;
+                }
+
+                update_data_result[partition_key].total_rows += data_block.rows();
+                it->second->write(data_block);
+                update_data_statistics.at(partition_key).update(chunk);
+            }
+        }
+
+        for (const auto & [partition_key, _] : update_data_result)
+        {
+            update_data_writers[partition_key]->flush();
+            update_data_writers[partition_key]->finalize();
+            update_data_write_buffers[partition_key]->finalize();
+            update_data_result[partition_key].total_bytes = static_cast<Int32>(update_data_write_buffers[partition_key]->count());
+        }
+    }
+
+    if (commands[0].type == MutationCommand::DELETE)
+        return WriteDataFilesResult{DeleteFileWriteResultWithStats{result, statistics}, std::nullopt};
+    else
+        return WriteDataFilesResult{DeleteFileWriteResultWithStats{result, statistics}, DeleteFileWriteResultWithStats{update_data_result, update_data_statistics}};
```

Comment:
I'd rename `result` to `delete_data_result`, for consistency with `update_data_result` and to make it more coherent. Same for `writers -> delete_data_writers` and `write_buffers -> delete_data_write_buffers`, `statistics -> delete_data_statistics`